core/src/main/java/org/elasticsearch/index/engine/Engine.java

```diff
@@ -592,25 +593,40 @@ protected static SegmentInfos readLastCommittedSegmentInfos(final ReferenceManag
      */
     public final SegmentsStats segmentsStats(boolean includeSegmentFileSizes) {
         ensureOpen();
-        try (Searcher searcher = acquireSearcher("segments_stats")) {
-            SegmentsStats stats = new SegmentsStats();
-            for (LeafReaderContext reader : searcher.reader().leaves()) {
-                final SegmentReader segmentReader = Lucene.segmentReader(reader.reader());
-                stats.add(1, segmentReader.ramBytesUsed());
-                stats.addTermsMemoryInBytes(guardedRamBytesUsed(segmentReader.getPostingsReader()));
-                stats.addStoredFieldsMemoryInBytes(guardedRamBytesUsed(segmentReader.getFieldsReader()));
-                stats.addTermVectorsMemoryInBytes(guardedRamBytesUsed(segmentReader.getTermVectorsReader()));
-                stats.addNormsMemoryInBytes(guardedRamBytesUsed(segmentReader.getNormsReader()));
-                stats.addPointsMemoryInBytes(guardedRamBytesUsed(segmentReader.getPointsReader()));
-                stats.addDocValuesMemoryInBytes(guardedRamBytesUsed(segmentReader.getDocValuesReader()));
-
-                if (includeSegmentFileSizes) {
-                    // TODO: consider moving this to StoreStats
-                    stats.addFileSizes(getSegmentFileSizes(segmentReader));
-                }
+        Set<String> segmentName = new HashSet<>();
+        SegmentsStats stats = new SegmentsStats();
+        try (Searcher searcher = acquireSearcher("segments_stats", SearcherScope.INTERNAL)) {
+            List<SegmentReader> internalSegmentReaders = searcher.reader().getContext()
+                .leaves().stream().map(l -> Lucene.segmentReader(l.reader())).collect(Collectors.toList());
+            internalSegmentReaders.forEach(r -> segmentName.add(r.getSegmentName()));
+            fillSegmentStats(includeSegmentFileSizes, internalSegmentReaders, stats);
+        }
+
+        try (Searcher searcher = acquireSearcher("segments_stats", SearcherScope.EXTERNAL)) {
+            Set<SegmentReader> externalSegmentReaders = searcher.reader().getContext()
+                .leaves().stream().map(l -> Lucene.segmentReader(l.reader()))
+                .filter(r -> segmentName.contains(r.getSegmentName()) == false)
+                .collect(Collectors.toSet());
+            fillSegmentStats(includeSegmentFileSizes, externalSegmentReaders, stats);
+        }
+        writerSegmentStats(stats);
+        return stats;
+    }
```

Comment:
I would find it easier to read without streams I think, by making `fillSegmentStats` take a single `SegmentReader` and then doing regular for loops over the segments of each index?